sql/handler.cc

```diff
@@ -1414,16 +1414,15 @@ int ha_commit_trans(THD *thd, bool all)
     goto err;
   }
 
-  if (rw_trans || thd->lex->sql_command == SQLCOM_ALTER_TABLE)
+  if (rw_trans && use_transaction_registry)
   {
-    if (use_transaction_registry && thd->vers_update_trt)
-    {
-      TR_table trt(thd, true);
-      if (trt.update())
-        goto err;
-      if (all)
-        commit_one_phase_2(thd, false, &thd->transaction.stmt, false);
-    }
+    TR_table trt(thd, true);
+    if (trt.update())
```

Comment:
This introduces redundant TRT open when there is no TRT data to push. Extra closing stmt transaction in this case breaks some tests.

Reply:
Thank you for pointing this out! I think that we need a different API: first query each participating storage engine, then invoke `trt.update(info)` if any storage engine indicated that it modified versioned data.